wallet-api/src/Wallet/API.hs

```diff
@@ -332,38 +334,43 @@ payToScript_ range addr v = void . payToScript range addr v
 
 -- | Collect all unspent outputs from a pay to script address and transfer them
 --   to a public key owned by us.
-collectFromScript :: (Monad m, WalletAPI m) => SlotRange -> ValidatorScript -> RedeemerScript -> m ()
-collectFromScript range scr red = do
-    am <- watchedAddresses
-    let addr = scriptAddress scr
-        outputs = am ^. at addr . to (Map.toList . fromMaybe Map.empty)
-        con (r, _) = scriptTxIn r scr red
-        ins        = con <$> outputs
-        value = fold $ fmap (txOutValue . snd) outputs
-
-    oo <- ownPubKeyTxOut value
-    void $ createTxAndSubmit range (Set.fromList ins) [oo]
+collectFromScript :: (WalletDiagnostics m, WalletAPI m) => SlotRange -> ValidatorScript -> RedeemerScript -> m ()
+collectFromScript = collectFromScriptFilter (\_ _ -> True)
 
 -- | Given the pay to script address of the 'ValidatorScript', collect from it
---   all the inputs that were produced by a specific transaction, using the
+--   all the outputs that were produced by a specific transaction, using the
 --   'RedeemerScript'.
 collectFromScriptTxn ::
-    (Monad m, WalletAPI m)
+    (WalletAPI m, WalletDiagnostics m)
     => SlotRange
     -> ValidatorScript
     -> RedeemerScript
     -> TxId
     -> m ()
-collectFromScriptTxn range vls red txid = do
+collectFromScriptTxn range vls red txid =
+    let flt k _ = txid == Ledger.txOutRefId k in
+    collectFromScriptFilter flt range vls red        
+
+-- | Given the pay to script address of the 'ValidatorScript', collect from it
+--   all the outputs that match a predicate, using the 'RedeemerScript'.
+collectFromScriptFilter :: 
+    (WalletAPI m, WalletDiagnostics m)
+    => (TxOutRef -> TxOut -> Bool)
+    -> SlotRange
+    -> ValidatorScript
+    -> RedeemerScript
+    -> m ()
+collectFromScriptFilter flt range vls red = do
     am <- watchedAddresses
     let adr     = Ledger.scriptAddress vls
         utxo    = fromMaybe Map.empty $ am ^. at adr
-        ourUtxo = Map.toList $ Map.filterWithKey (\k _ -> txid == Ledger.txOutRefId k) utxo
+        ourUtxo = Map.toList $ Map.filterWithKey flt utxo
         i ref = scriptTxIn ref vls red
         inputs = Set.fromList $ i . fst <$> ourUtxo
         value  = fold $ fmap (txOutValue . snd) ourUtxo
 
     out <- ownPubKeyTxOut value
+    warnEmptyTransaction value adr
```

Comment:
Alternatively: we could warn in `createTxAndSubmit` if the value in the tx is 0, or even if any input has zero value.

Reply:
In `createTxAndSubmit` we only have the transaction inputs. They don't have a `Value` so we can't compute the total value of all inputs consumed by the transaction. We could use the total value of outputs & the fee instead, but with the multi currency ledger's ability to destroy currency it is not inconceivable that one wants to submit a transaction with zero outputs. 

`collectFromScript` is a higher-level API for a common use case so I think it's ok to just put the warning there.

Reply:
Fair